lib/vector-common/src/shutdown.rs

```diff
@@ -107,9 +107,11 @@ impl ShutdownSignal {
     }
 }
 
+type IsInternal = bool;
+
 #[derive(Debug, Default)]
 pub struct SourceShutdownCoordinator {
-    shutdown_begun_triggers: HashMap<ComponentKey, Trigger>,
+    shutdown_begun_triggers: HashMap<ComponentKey, (IsInternal, Trigger)>,
```

Comment:
Did you consider the approach of having two shutdown sets, and handling them separately?

Reply:
Hey! I tried separating concerns between internal and external triggers, but I think that it adds a bit more complexity for the reader: 
- it's longer
- we have to duplicate some logic or fall back to inefficient try-and-fail patterns

Here is the commit with the change: https://github.com/vectordotdev/vector/commit/7d7c51f82025525acbf171fe179063288900a8c2

Let me know if you want me to merge it to this branch! 🚀

Reply:
No, that looks substantially worse than this approach, so I don't think it's worth it, at least not at this point. Leave it this way.